### PR TITLE
Simplify cache namespace and key encoding logic for v1

### PR DESCRIPTION
**Pull request by pshafer-als**
## What do these changes do?

* All logic to select the namespace for a cache key, and encode the resulting namespaced key are now encapsulated in the `build_key()` member of `BaseCache` and its backend subclasses.
* Simplifies the burden of creating a custom ``key_builder`` for a cache.
* Consistent use of namespaced keys across backends and helpers (e.g., locks) while still allowing customization of the formatting and encoding of the namespaced key.

## Are there changes in behavior for the user?

* When creating a cache object with a custom ``key_builder`` argument...
    * You no longer need to consider the backend encoding requirements for key identifiers; these are handled by the backend cache class.
    * The ``namespace`` argument passed to your custom ``key_builder`` has already been resolved by the cache: It will be the namespace that was explicitly requested by a client (if requested); otherwise it will be the default namespace that was used to instantiate the cache.
* If your code was calling the private `_build_key()` member of a cache, replace those calls with `build_key()`.

## Related issue number

Resolves issue #669.

## Checklist

- [X ] I think the code is well written
- [X ] Unit tests for the changes exist
- [X ] Documentation reflects the changes
- [X ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

File: aiocache/backends/redis.py [hunk 1]
```diff
@@ -218,14 +218,8 @@ def parse_uri_path(cls, path):
...
-    def _build_key(self, key, namespace=None):
-        if namespace is not None:
-            return "{}{}{}".format(
-                namespace, ":" if namespace else "", _ensure_key(key))
-        if self.namespace is not None:
-            return "{}{}{}".format(
-                self.namespace, ":" if self.namespace else "", _ensure_key(key))
-        return key
+    def _build_key_default(self, key, namespace=None):
+        return "{}{}{}".format(namespace or "", ":" if namespace else "", key)
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:222]: If we change to always `str`, then this becomes:
```suggestion
        return "{}:{}".format(namespace, key) if namespace else key
```

File: aiocache/base.py [hunk 2]
```diff
@@ -498,12 +489,13 @@ async def close(self, *args, _conn=None, **kwargs):
...
-            return "{}{}".format(namespace, _ensure_key(key))
-        if self.namespace is not None:
-            return "{}{}".format(self.namespace, _ensure_key(key))
-        return key
+    def build_key(self, key, namespace=None):
+        key_name = key.value if isinstance(key, Enum) else key
+        ns = namespace if namespace is not None else self.namespace
+        return self._build_key(key_name, namespace=ns)
+
+    def _build_key_default(self, key, namespace=None):
```
Comment (Dreamsorcerer) [inline aiocache/base.py:497]: I'm still inclined to just pass a lambda as the default value in `__init__()`.

**Comment by Dreamsorcerer** [comment]
Would also be nice to add type annotations to the methods that have been edited/added (and the key_builder parameter), to get some movement in that direction.

**Comment by pshafer-als** [comment]
> Would also be nice to add type annotations to the methods that have been edited/added (and the key_builder parameter), to get some movement in that direction.

Sure.  That should be a simple improvement.

File: aiocache/base.py [hunk 3]
```diff
@@ -498,12 +489,13 @@ async def close(self, *args, _conn=None, **kwargs):
...
+        return "{}{}".format(namespace or "", key)
```
[the remark this answers is not in the document]
Comment (pshafer-als) [reply aiocache/base.py:498]: This difference is meaningful at the level above this point (`BaseCache.build_key()`) because `namespace=None` will use the cache's default, whereas `namespace=""` is an override of the cache's default to use no namespace prefix.

I think you're right that at this level (`BaseCache._build_key()`) there is _**probably**_ no difference between `None` and `""`.  However, that does take away that particular freedom from the end user who might care about that distinction in their custom `key_builder`.

My inclination is slightly in favor of keeping this as is, but it's only based on _possibility_ rather than a concrete use case.

File: aiocache/backends/redis.py (hunk 1 above)
Comment (pshafer-als) [reply aiocache/backends/redis.py:222]: Right

File: aiocache/base.py (hunk 2 above)
Comment (pshafer-als) [reply aiocache/base.py:497]: Yeah, maybe you're right.  I had thought that this could provide a nice 'template' for end users who write their own key_builder; and it's clear when it gets overridden (like in the redis backend).

But if you think that the benefit is minimal, I'm ok with using the lambda.

File: aiocache/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Dreamsorcerer) [reply aiocache/base.py:498]: Yep, but I can't think of any real cases for it, and ensuring it is always `str` means that the key builders all become simpler code and less likely that a user writes incorrect code (or has to special case `None` even though they will never use it).

[the remark this answers is not in the document]
Comment (pshafer-als) [reply aiocache/base.py:498]: That makes sense. It might be the tiniest smidge faster also.

Let's give it a try.

File: aiocache/base.py [hunk 4]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
     NAME: str
 
     def __init__(
-        self, serializer=None, plugins=None, namespace=None, key_builder=None, timeout=5, ttl=None
+        self,
+        serializer: Optional[BaseSerializer] = None,
+        plugins: Optional[List[Any]] = None,  # aiocache.plugins depends on aiocache.base
```
Comment (pshafer-als) [inline aiocache/base.py:115]: Using `Any` because `aiocache.plugins` depends on `aiocache.base`.

I tried [refactoring `API` into its own module](https://github.com/pshafer-als/aiocache/blob/simplify-cache-namespace-refactor-api/aiocache/api.py), but it causes a lot of tests to fail.  At the moment I'm presuming this is `@API.register` writing to the class variables of different objects

File: aiocache/base.py [hunk 5]
```diff
@@ -498,12 +498,13 @@ async def close(self, *args, _conn=None, **kwargs):
...
-        if self.namespace is not None:
-            return "{}{}".format(self.namespace, _ensure_key(key))
-        return key
+    def build_key(self, key: str, namespace: Optional[str] = None) -> str:
+        key_name = key.value if isinstance(key, Enum) else key
+        ns = namespace if namespace is not None else self.namespace
+        return self._build_key(key_name, namespace=ns)
+
+    def _build_key_default(self, key: str, namespace: str = "") -> str:
+        return f"{namespace}{key}"
```
Comment (pshafer-als) [inline aiocache/base.py:507]: Will delete this in the next commit.

**Comment by pshafer-als** [comment]
I added the changes we discussed, and started on the type annotations.  I could continue adding type annotations in these files, but would like to get your feedback before proceeding further.

File: aiocache/base.py (hunk 4 above)
Comment (pshafer-als) [reply aiocache/base.py:115]: > At the moment I'm presuming this is @API.register writing to the class variables of different objects

That wasn't the problem.  It's an 'order of import' problem, where `aiocache.plugins` is attempting to create hooks before `aiocache.base` registers any commands.

Another approach would be to move the base classes into a separate model, so that they can be used for type annotations.  I'll give that a try to see how 'forced' it feels.

Comment (pshafer-als) [reply aiocache/base.py:115]: > Another approach would be to move the base classes into a separate module, so that they can be used for type annotations.

Yes, [that approach is cleaner](https://github.com/pshafer-als/aiocache/blob/simplify-cache-namespace-with-types/aiocache/types.py). If this becomes repeatedly necessary as type annotations are added to other parts of the codebase, then we could move other base classes into the 'types' module as needed.

File: aiocache/base.py [hunk 6]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
...
-        self._serializer = None
-        self.serializer = serializer or serializers.StringSerializer()
+        self.namespace: str = namespace
+        self._build_key: Callable[[str, str], str] = key_builder or (
+            lambda key, namespace="": f"{namespace}{key}")
 
-        self._plugins = None
-        self.plugins = plugins or []
+        self._serializer: Optional[BaseSerializer] = None
+        self.serializer: BaseSerializer = serializer or StringSerializer()
```
Comment (pshafer-als) [inline aiocache/base.py:129]: Here and elsewhere, should we replace the base class with a `TypeVar`?

Something like...
```
from typing import TypeVar
SerializerType = TypeVar('SerializerType', bound='BaseSerializer')

self.serializer: SerializerType = serializer or StringSerializer()
```

**Comment by Dreamsorcerer** [comment]
> I added the changes we discussed, and started on the type annotations. I could continue adding type annotations in these files, but would like to get your feedback before proceeding further.

Let's keep the changes to a minimum for this PR, further type annotations can be done in another PR. However, I suspect that completing some of the 1.0 refactoring tasks first will make it easier.

File: aiocache/backends/memcached.py [hunk 7]
```diff
@@ -147,8 +148,8 @@ def __init__(self, serializer=None, **kwargs):
     def parse_uri_path(cls, path):
         return {}
 
-    def _build_key(self, key, namespace=None):
-        ns_key = super()._build_key(key, namespace=namespace).replace(" ", "_")
+    def build_key(self, key: str, namespace: Optional[str] = None) -> str:
```
Comment (Dreamsorcerer) [inline aiocache/backends/memcached.py:151]: Should probably try to diff the mypy results to see if there are new errors. master currently has 20 errors:
https://github.com/aio-libs/aiocache/actions/runs/4235473010/jobs/7359129847

This must be an error, because the function returns `bytes`.

Maybe we need a `TypeVar` to declare the type here (which can be defined in each subclass).

File: aiocache/backends/redis.py [hunk 8]
```diff
@@ -218,14 +230,8 @@ def parse_uri_path(cls, path):
...
-    def _build_key(self, key, namespace=None):
-        if namespace is not None:
-            return "{}{}{}".format(
-                namespace, ":" if namespace else "", _ensure_key(key))
-        if self.namespace is not None:
-            return "{}{}{}".format(
-                self.namespace, ":" if self.namespace else "", _ensure_key(key))
-        return key
+    def _build_key_default(self, key: str, namespace: str = "") -> str:
+        return f"{namespace}:{key}" if namespace else key
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:234]:
```suggestion
```

File: aiocache/backends/redis.py [hunk 9]
```diff
@@ -199,8 +200,19 @@ class RedisCache(RedisBackend):
...
+        serializer: Optional[BaseSerializer] = None,
+        namespace: str = "",
+        key_builder: Optional[Callable[[str, str], str]] = None,
+        **kwargs,
+    ):
+        super().__init__(
+            serializer=serializer or JsonSerializer(),
+            namespace=namespace,
+            key_builder=key_builder or (
+                lambda key, namespace="": f"{namespace}:{key}" if namespace else key),
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:214]: Default is unneeded, we always pass namespace in build_key().
```suggestion
                lambda key, namespace: f"{namespace}:{key}" if namespace else key),
```

File: aiocache/base.py (hunk 4 above)
Comment (Dreamsorcerer) [reply aiocache/base.py:115]: For now, just import it with `if TYPE_CHECKING:`. Further refactoring can be considered in another PR if you still feel it's appropriate.

File: aiocache/base.py [hunk 10]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
...
+        key_builder: Optional[Callable[[str, str], str]] = None,
+        timeout: Optional[float] = 5,
+        ttl: Optional[float] = None,
     ):
-        self.timeout = float(timeout) if timeout is not None else timeout
-        self.namespace = namespace
-        self.ttl = float(ttl) if ttl is not None else ttl
-        self.build_key = key_builder or self._build_key
+        self.timeout: Optional[float] = float(timeout) if timeout is not None else None
+        self.ttl: Optional[float] = float(ttl) if ttl is not None else None
```
Comment (Dreamsorcerer) [inline aiocache/base.py:122]:
```suggestion
        self.timeout = float(timeout) if timeout is not None else None
        self.ttl = float(ttl) if ttl is not None else None
```

Double check with `reveal_type`, but mypy should infer the types for these attributes.

File: aiocache/base.py [hunk 11]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
...
 
-        self._serializer = None
-        self.serializer = serializer or serializers.StringSerializer()
+        self.namespace: str = namespace
+        self._build_key: Callable[[str, str], str] = key_builder or (
```
Comment (Dreamsorcerer) [inline aiocache/base.py:125]:
```suggestion
        self._build_key = key_builder or (
```
Hopefully this one too.

File: aiocache/base.py [hunk 12]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
...
+            lambda key, namespace="": f"{namespace}{key}")
```
Comment (Dreamsorcerer) [inline aiocache/base.py:126]:
```suggestion
            lambda key, namespace: f"{namespace}{key}")
```

File: aiocache/base.py (hunk 6 above)
Comment (Dreamsorcerer) [reply aiocache/base.py:129]: I'm not sure what benefit that brings. But, leave it to a later PR if you think there is a benefit.

File: aiocache/base.py [hunk 13]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
...
 
-        self._plugins = None
-        self.plugins = plugins or []
+        self._serializer: Optional[BaseSerializer] = None
+        self.serializer: BaseSerializer = serializer or StringSerializer()
+
+        self._plugins: List[Any] = None
+        self.plugins: List[Any] = plugins or []
```
Comment (Dreamsorcerer) [inline aiocache/base.py:132]:
```suggestion
        self._serializer = None
        self.serializer = serializer or StringSerializer()

        self._plugins: Optional[List[BasePluign]] = None
        self.plugins = plugins or []
```

Not sure what I'm missing, but `_plugins` doesn't make much sense. I'm not sure why it's initialised to `None` or where it later gets set to something else.

Comment (Dreamsorcerer) [reply aiocache/base.py:132]: `plugins` is also a property method, so that's broken too... Maybe someone meant to do `_plugins = plugins or []` and then use the property method?

File: aiocache/base.py [hunk 14]
```diff
@@ -418,8 +419,7 @@ async def expire(self, key, ttl, namespace=None, _conn=None):
         :raises: :class:`asyncio.TimeoutError` if it lasts more than self.timeout
         """
         start = time.monotonic()
-        ns = namespace if namespace is not None else self.namespace
-        ns_key = self.build_key(key, namespace=ns)
+        ns_key = self.build_key(key, namespace=namespace)
```
Comment (Dreamsorcerer) [inline aiocache/base.py:422]: I'd be tempted to remove the kwarg from all these, rather than repeating the name.

**Comment by pshafer-als** [comment]
Later today I'll clean this up per your recommendations.  After that I'm thinking it might make sense in this PR to replace `namespace=None` with `namespace=""` in the call signature for all the BaseCache member functions.

e.g.
https://github.com/aio-libs/aiocache/blob/db9426196d0a4b55a78411c8b96949af645a557b/aiocache/base.py#L154

What do you think?

File: aiocache/base.py [hunk 15]
```diff
@@ -498,12 +498,13 @@ async def close(self, *args, _conn=None, **kwargs):
...
-    def _build_key(self, key, namespace=None):
-        if namespace is not None:
-            return "{}{}".format(namespace, _ensure_key(key))
-        if self.namespace is not None:
-            return "{}{}".format(self.namespace, _ensure_key(key))
-        return key
+    def build_key(self, key: str, namespace: Optional[str] = None) -> str:
+        key_name = key.value if isinstance(key, Enum) else key
+        ns = namespace if namespace is not None else self.namespace
+        return self._build_key(key_name, namespace=ns)
```
Comment (Dreamsorcerer) [inline aiocache/base.py:504]:
```suggestion
        return self._build_key(key_name, ns)
```
Our typing guarantees 2 positional arguments, it doesn't guarantee the keyword name.

File: tests/ut/test_base.py [hunk 16]
```diff
@@ -218,16 +218,15 @@ def test_alt_build_key(self):
     def alt_base_cache(self, init_namespace="test"):
         """Custom key_builder for cache"""
         def build_key(key, namespace=None):
-            ns = namespace if namespace is not None else ""
             sep = ":" if namespace else ""
-            return f"{ns}{sep}{_ensure_key(key)}"
+            return f'{namespace or ""}{sep}{ensure_key(key)}'
```
Comment (Dreamsorcerer) [inline tests/ut/test_base.py:222]:
```suggestion
            return f'{namespace}{sep}{ensure_key(key)}'
```

**Comment by Dreamsorcerer** [comment]
> Later today I'll clean this up per your recommendations. After that I'm thinking it might make sense in this PR to replace `namespace=None` with `namespace=""` in the call signature for all the BaseCache member functions.
> 
> e.g.
> 
> https://github.com/aio-libs/aiocache/blob/db9426196d0a4b55a78411c8b96949af645a557b/aiocache/base.py#L154
> 
> What do you think?

Wouldn't that break the default namespace behaviour? i.e. It needs to pass `None` to `build_key()` to use the default namespace.

File: aiocache/base.py [hunk 17]
```diff
@@ -498,12 +498,13 @@ async def close(self, *args, _conn=None, **kwargs):
...
 
-    def _build_key(self, key, namespace=None):
-        if namespace is not None:
-            return "{}{}".format(namespace, _ensure_key(key))
-        if self.namespace is not None:
-            return "{}{}".format(self.namespace, _ensure_key(key))
-        return key
+    def build_key(self, key: str, namespace: Optional[str] = None) -> str:
+        key_name = key.value if isinstance(key, Enum) else key
+        ns = namespace if namespace is not None else self.namespace
```
Comment (Dreamsorcerer) [inline aiocache/base.py:503]:
```suggestion
        ns = self.namespace if namespace is None else namespace
```
I'd always try to do the positive test.

**Comment by pshafer-als** [comment]
> > After that I'm thinking it might make sense in this PR to replace `namespace=None` with `namespace=""` in the call signature for all the BaseCache member functions.
>
> Wouldn't that break the default namespace behaviour? i.e. It needs to pass `None` to `build_key()` to use the default namespace.

Er, yeah it would.  I came back here to say the same thing, but you beat me to it.  :)

File: aiocache/backends/memcached.py (hunk 7 above)
Comment (Dreamsorcerer) [reply aiocache/backends/memcached.py:151]: Or 12 errors after installing all missing libraries:
https://github.com/aio-libs/aiocache/actions/runs/4235831591/jobs/7359931126

File: aiocache/base.py (hunk 6 above)
Comment (pshafer-als) [reply aiocache/base.py:129]: That's to indicate that the type can be a subclass of the bound type; so `StringSerializer` is not expected to be coerced to `BaseSerializer`, for example.

> But, leave it to a later PR if you think there is a benefit.

Agreed.

File: aiocache/backends/memcached.py (hunk 7 above)
Comment (pshafer-als) [reply aiocache/backends/memcached.py:151]: To use a `TypeVar` in the return type, we'll also need either:
1. The same `TypeVar` to annotate on the function parameters.  We could add a third, otherwise unused, parameter for this purpose, but that seems pretty kludged.
2. Derive `BaseCache` from `Generic[]`.  That also seems heavy handed, just for type checking.

Using a simple `Union[str, bytes]` will work as a return type for all existing cases. This could be done with a type alias at the module level. Trying to add the alias within `BaseCache` and its subclasses requires the `TypeAlias` annotation, which is only available for python >= 3.10.

I'll see whether this approach plays well with mypy.

File: aiocache/base.py (hunk 11 above)
Comment (pshafer-als) [reply aiocache/base.py:125]: Mypy is fine without this annotation, but it might be useful for documenting the code?  This is because lambdas can't have type annotations.

File: aiocache/base.py (hunk 13 above)
Comment (pshafer-als) [reply aiocache/base.py:132]: You're right.  This is strange.  I'll reduce these two initialization assignments into one, for `_serializer` and for `_plugins`.

`self.serializer` and `self._serializer` are used interchangeably throughout this module.  `self.serializer` would be appropriate in all cases.

---
Although the getter/setter don't do anything useful now, they do provide a mechanism for subclasses to do any customizations they might need.  I'd like to defer removing them to a future PR.

File: aiocache/base.py (hunk 15 above)
Comment (pshafer-als) [reply aiocache/base.py:504]: Good point.

File: aiocache/base.py [hunk 18]
```diff
@@ -109,18 +110,26 @@ class BaseCache:
     NAME: str
 
     def __init__(
-        self, serializer=None, plugins=None, namespace=None, key_builder=None, timeout=5, ttl=None
+        self,
+        serializer: Optional[BaseSerializer] = None,
+        plugins: Optional[List[Any]] = None,  # aiocache.plugins depends on aiocache.base
+        namespace: str = "",
```
Comment (pshafer-als) [inline aiocache/base.py:116]: This breaks `build_key()` if `None` is passed as an argument here. 

Either we allow `None` here and convert it to `""` inside `__init__()`, or we should add a test that expects to fail when `None` is passed here.

Which would you recommend?

File: aiocache/base.py [hunk 19]
```diff
@@ -87,6 +91,7 @@ async def _plugins(self, *args, **kwargs):
         return _plugins
 
 
+# class BaseCache(Generic[CacheKeyType]):
```
Comment (pshafer-als) [inline aiocache/base.py:94]: Here and elsewhere in comments are how we might implement typing of the `build_key()` return type if we decide using Generics would be more useful or robust.

Comment (pshafer-als) [reply aiocache/base.py:94]: It probably means also updating tests that check the call signature of `__init()__`, and maybe a few others.

**Comment by pshafer-als** [comment]
@Dreamsorcerer I'll need to pull in the latest dependency bumps, and update at least one failing test still.  I'm pausing now to share the current state of the PR for your feedback.

File: aiocache/backends/memcached.py [hunk 20]
```diff
@@ -139,16 +142,22 @@ class MemcachedCache(MemcachedBackend):
...
         return {}
 
-    def _build_key(self, key, namespace=None):
-        ns_key = super()._build_key(key, namespace=namespace).replace(" ", "_")
+    # def build_key(self, key: str, namespace: Optional[str] = None) -> CacheKeyType:
+    #     ns_key = super().build_key(key, namespace=namespace).replace(" ", "_")
+    #     return str.encode(ns_key)
+
+    def build_key(self, key: str, namespace: Optional[str] = None) -> KeyType:
+        ns_key = cast(str, super().build_key(key, namespace=namespace)).replace(" ", "_")
```
Comment (pshafer-als) [inline aiocache/backends/memcached.py:160]: An alternative to `cast` could simply be `# type: ignore`

File: aiocache/backends/memcached.py (hunk 7 above)
Comment (Dreamsorcerer) [reply aiocache/backends/memcached.py]: Yes, I meant with a `Generic`, the functions for `_get()` etc. would all need to use the same TypeVar for the key parameter. This should then type check correctly when you pass the key to memcached which expects bytes (MemcacheBackend would subclass `BaseCache[bytes]`).

File: aiocache/base.py (hunk 18 above)
Comment (Dreamsorcerer) [reply aiocache/base.py:121]: A runtime check? Sure, that's fine.

I'd vote for not allowing `None` though and keeping the code simpler.

File: aiocache/base.py (hunk 19 above)
Comment (Dreamsorcerer) [reply aiocache/base.py:94]: I don't think anything needs to be changed to `__init__()`.

File: aiocache/backends/memcached.py (hunk 20 above)
Comment (Dreamsorcerer) [reply aiocache/backends/memcached.py:160]: Hmm, yes, that's slightly awkward.

Maybe a type safe option would be to make the `BaseCache` define it as an abstract method, and provide this default method under another name (like _str_build_key() or something)?

Comment (pshafer-als) [reply aiocache/backends/memcached.py:160]: > Maybe a type safe option would be to make the `BaseCache` define it as an abstract method, and provide this default method under another name (like `_str_build_key()` or something)?

I think you're right.  I've been trying to make this work by overloading `build_key()` with `@singledispatchmethod` to no avail.  For that I would need to implement some factory function that returns a `BaseCache` with the appropriate return type for `build_key()`.  Implementing this private helper method seems much simpler.

File: aiocache/base.py (hunk 19 above)
Comment (pshafer-als) [reply aiocache/base.py:94]: I was referring to this potential change [_below_], which was intended as a convenience...providing the default`build_key()` return type of `str` so that it could be inferred by mypy, even when not explicitly created with `BaseCache[str]`.

https://github.com/aio-libs/aiocache/blob/84ce34c07b1e2961583c9f3beeb2c146da2ba3d4/aiocache/base.py#L118-L126
[_although the last line should have been `key_type:Type[CacheKeyType] = str`_]

But, unfortunately, it's not (yet) possible to supply a [default type for the `TypeVar`](https://github.com/python/mypy/issues/4236). Oh well.

Comment (Dreamsorcerer) [reply aiocache/base.py:94]: You can use overloads to achieve that, but I don't think it's relevant. There should be no parameter here, all subclasses should explicitly define the type they use.

File: aiocache/base.py (hunk 14 above)
Comment (pshafer-als) [reply aiocache/base.py:427]: Generally I like to call with keyword args if I think the function signature is likely to change (and of course if some kwargs with defaults are skipped).

But in this case I think it's fine use the positional arguments as you've suggested.  The argument name at least conveys the intent.

File: aiocache/base.py (hunk 11 above)
Comment (pshafer-als) [reply aiocache/base.py:132]: > Frankly, you could just define this lambda as the parameter default and disallow `None` completely.

I like this suggestion, but mypy does not. :(

File: aiocache/backends/memcached.py (hunk 20 above)
Comment (pshafer-als) [reply aiocache/backends/memcached.py]: `build_key()` is now an abstract method on `BaseCache`; and overloaded/defined for the derived classes.

File: aiocache/base.py (hunk 19 above)
Comment (pshafer-als) [reply aiocache/base.py]: The extra `__init__()` parameter for implicit typing idea has been scrapped.

File: aiocache/base.py (hunk 18 above)
Comment (pshafer-als) [reply aiocache/base.py:120]: For now, have omitted a runtime check.  Passing `namespace=None` does not cause an error; instead it prepends the "None" namespace to the key.

This is probably unexpected behavior, but doesn't cause any real problem.  I could add a runtime warning to alert the user that `""` (or no argument) is the expected value for a cache constructor.

File: aiocache/base.py (hunk 11 above)
Comment (Dreamsorcerer) [reply aiocache/base.py:129]: type: ignore it if it's a bug in mypy.

I know there's this issue around lambdas already: https://github.com/python/mypy/issues/9590

File: aiocache/backends/memcached.py [hunk 21]
```diff
@@ -119,6 +120,7 @@ async def _close(self, *args, _conn=None, **kwargs):
         await self.client.close()
 
 
+# class MemcachedCache(MemcachedBackend):
```
Comment (Dreamsorcerer) [inline aiocache/backends/memcached.py:123]:
```suggestion
```

File: aiocache/base.py (hunk 11 above)
Comment (pshafer-als) [reply aiocache/base.py:129]: It turns out this was my error.  I think I must have missed a comma at the end of the line or something mundane like that.  Correcting it now.  Will push in the next round of updates for this PR.

File: aiocache/backends/memory.py [hunk 22]
```diff
@@ -1,17 +1,21 @@
...
     """
     Wrapper around dict operations to use it as a cache backend
     """
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
+        # Assigning build_key here is clear and avoids the extra stack frame
+        # from nesting the call to _str_build_key() in an override of build_key()
+        # ...but mypy needs the override definition to recognize a concrete class
+        # self.build_key = self._str_build_key  # Simple, but not mypy-friendly
```
Comment (Dreamsorcerer) [inline aiocache/backends/memory.py:18]:
```suggestion
```

Not just mypy, I'd expect it to fail at runtime too if it's abstract.

File: aiocache/backends/memory.py [hunk 23]
```diff
@@ -131,3 +135,6 @@ def __init__(self, serializer=None, **kwargs):
     @classmethod
     def parse_uri_path(cls, path):
         return {}
+
+    def build_key(self, key: str, namespace: Optional[str] = None) -> str:
+        return self._str_build_key(key, namespace)
```
Comment (Dreamsorcerer) [inline aiocache/backends/memory.py:140]: You might, however, be able to just do:
`build_key = _str_build_key`
Rather than trying to set it in the `__init__()` method.

File: aiocache/base.py [hunk 24]
```diff
@@ -3,16 +3,21 @@
...
+if TYPE_CHECKING:
+    from aiocache.plugins import BasePlugin
+    from aiocache.serializers import BaseSerializer
+from aiocache.serializers import StringSerializer
 
 
 logger = logging.getLogger(__name__)
 
 SENTINEL = object()
+CacheKeyType = TypeVar("CacheKeyType", bound=Union[str, bytes])
```
Comment (Dreamsorcerer) [inline aiocache/base.py:20]: I think it's totally up to the class what type they use for cache keys, so I'd leave this unbounded.

Maybe some special cache uses `int` after producing some kind of hash?

Comment (pshafer-als) [reply aiocache/base.py:20]: OK, now that we use the `_str_build_key()` helper, this bound is more restrictive than helpful.

File: aiocache/factory.py [hunk 25]
```diff
@@ -172,12 +172,12 @@ def get(self, alias: str) -> object:
...
         return cache
 
-    def create(self, alias: str, **kwargs):
+    def create(self, alias, **kwargs):
         """Create a new cache.
 
         You can use kwargs to pass extra parameters to configure the cache.
 
-        :param alias: alias to pull configuration from
+        :param alias (str): alias to pull configuration from
```
Comment (Dreamsorcerer) [inline aiocache/factory.py:180]: Not sure why this file has been changed, seems unrelated.

Comment (pshafer-als) [reply aiocache/factory.py:180]: It was a mypy error.  Should I revert it and submit a separate PR?

File: tests/performance/test_footprint.py [hunk 26]
```diff
@@ -1,13 +1,15 @@
...
+from typing import Any, AsyncGenerator
 
 import aiomcache
 import pytest
 import redis.asyncio as redis
 
 
 @pytest.fixture
-async def redis_client() -> redis.Redis:
+async def redis_client() -> "AsyncGenerator[redis.Redis[Any], None]":
```
Comment (Dreamsorcerer) [inline tests/performance/test_footprint.py:11]:
```suggestion
async def redis_client() -> "AsyncIterator[redis.Redis[Any]]":
```

File: tests/performance/test_footprint.py [hunk 27]
```diff
@@ -1,13 +1,15 @@
...
+    r: "redis.Redis[Any]"
```
Comment (Dreamsorcerer) [inline tests/performance/test_footprint.py:12]: Not clear why this is needed.

Comment (pshafer-als) [reply tests/performance/test_footprint.py:12]: It was a mypy error.

File: tests/ut/conftest.py [hunk 28]
```diff
@@ -29,27 +29,39 @@ def reset_caches():
...
             stack.enter_context(patch.object(cache, f, autospec=True))
         stack.enter_context(patch.object(cache, "_serializer", autospec=True))
+        stack.enter_context(patch.object(cache, "build_key", cache._str_build_key))
         yield cache
 
 
+@pytest.fixture
+def generic_base_cache():
+    # TODO: Is there need for a separate BaseCache[bytes] fixture?
+    return BaseCache[str]()
```
Comment (Dreamsorcerer) [inline tests/ut/conftest.py:56]: Also, task for another PR. BaseCache should be ABC, which means it then won't be able to be instantiated directly. This should probably then just use the memory cache for testing or something..

File: tests/ut/test_decorators.py [hunk 29]
```diff
@@ -60,7 +60,7 @@ def test_fails_at_instantiation(self):
         with pytest.raises(TypeError):
 
             @cached(wrong_param=1)
-            async def fn() -> None:
+            async def fn():
```
Comment (Dreamsorcerer) [inline tests/ut/test_decorators.py:63]: This file looks unrelated again.

File: aiocache/backends/memory.py (hunk 23 above)
Comment (pshafer-als) [reply aiocache/backends/memory.py:140]: I'm having no luck with that approach.

File: aiocache/backends/memory.py (hunk 22 above)
Comment (Dreamsorcerer) [reply aiocache/backends/memory.py:18]: That's because it's not actually an ABC, which it should be (later PR though).

File: aiocache/factory.py (hunk 25 above)
Comment (Dreamsorcerer) [reply aiocache/factory.py:180]: Ignore the existing 12 errors:
https://github.com/aio-libs/aiocache/actions/runs/4235831591/jobs/7359931126

They can be dealt with separately. I think we just want to remove this entire file, as I don't see what benefit this part adds (but makes the code unnecessarily complex and fail type checking).

Comment (pshafer-als) [reply aiocache/factory.py:180]: OK, I'll remove the extraneous mypy fixes from this PR.

File: tests/ut/test_decorators.py (hunk 29 above)
Comment (pshafer-als) [reply tests/ut/test_decorators.py]: Reverted these changes

File: tests/ut/conftest.py (hunk 28 above)
Comment (pshafer-als) [reply tests/ut/conftest.py:54]: Sounds good.

File: tests/performance/test_footprint.py (hunk 27 above)
Comment (pshafer-als) [reply tests/performance/test_footprint.py]: Reverted these changes

File: tests/performance/test_footprint.py (hunk 26 above)
Comment (pshafer-als) [reply tests/performance/test_footprint.py]: Done (and then reverted per subsequent comment)

File: aiocache/base.py (hunk 13 above)
Comment (pshafer-als) [reply aiocache/base.py]: Will leave any getter/setter changes to a subsequent PR.

**Comment by Dreamsorcerer** [comment]
If you want to work on any more, I think #676 and #625 are probably good next issues.

**Comment by Dreamsorcerer** [comment]
#677 as well. Should be easier to get some typing working then.
